cert_manager/generate_cert.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Provides utility methods to generate X.509 certificates with different
+// options. This implementation is Largely inspired from
+// https://golang.org/src/crypto/tls/generate_cert.go.
+
+package cert_manager
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"io/ioutil"
+	"log"
+	"math/big"
+	"net"
+	"strings"
+	"time"
+)
+
+type CertOptions struct {
+	// Comma-separated hostnames and IPs to generate a certificate for.
+	// This can also be set to the identity running the workload,
+	// like kubernetes service account.
+	Host string
+
+	// Creation date formatted as Jan 1 15:04:05 2011.
+	ValidFrom string
+
+	// Duration that certificate is valid for.
+	ValidFor time.Duration
+
+	// Signer private key file (PEM encoded).
+	SignerPriv string
+
+	// Signer certificate file (PEM encoded).
+	SignerCert string
+
+	// Juju org", "Organization for the cert.
+	Org string
+
+	// Whether this cert should be a Cerificate Authority.
+	IsCA bool
+
+	// Whether this cerificate is self-signed.
+	IsSelfSigned bool
+
+	// Whether this certificate is for a client.
+	IsClient bool
+}
+
+// Size of RSA key to generate.
+const RsaBits = 2048
+
+// GenCert generates a X.509 certificate with the given options.
+func GenCert(options CertOptions) ([]byte, []byte) {
+	// Generates a RSA private&public key pair.
+	// The public key will be bound to the certficate generated below. The
+	// private key will be used to sign this certificate in the self-signed
+	// case, otherwise the certificate is signed by the signer private key
+	// as specified in the CertOptions.
+	priv, err := rsa.GenerateKey(rand.Reader, RsaBits)
+	if err != nil {
+		log.Fatalf("RSA key generation failed with error %s.", err)
+	}
+	template := genCertTemplate(options)
+	signingCert, signingKey := &template, priv
+	if !options.IsSelfSigned {
+		signingCert, signingKey = loadSigningCreds(options.SignerPriv, options.SignerCert)
```

Comment:
It feels strange that the signing key and cert are passed in by the caller. Instead I think the CA should generate/read from a file during startup and use the same key & cert to sign all requests.